File: solutions/057-insert-interval/solution.py
from typing import List


class Solution:
    def insert(
        self, intervals: List[List[int]], new_interval: List[int]
    ) -> List[List[int]]:
        merged = [new_interval[0], new_interval[1]]
        result: list[list[int]] = []
        i, n = 0, len(intervals)

        while i < n and intervals[i][1] < merged[0]:
            result.append(intervals[i])
            i += 1

        while i < n and intervals[i][0] <= merged[1]:
            merged[0] = min(merged[0], intervals[i][0])
            merged[1] = max(merged[1], intervals[i][1])
            i += 1
        result.append(merged)

        while i < n:
            result.append(intervals[i])
            i += 1

        return result


if __name__ == "__main__":
    tests = [
        ([[1, 3], [6, 9]], [2, 5], [[1, 5], [6, 9]]),
        (
            [[1, 2], [3, 5], [6, 7], [8, 10], [12, 16]],
            [4, 8],
            [[1, 2], [3, 10], [12, 16]],
        ),
        ([], [5, 7], [[5, 7]]),
        ([[1, 5]], [2, 3], [[1, 5]]),
    ]
    sol = Solution()
    for intervals, new_interval, expected in tests:
        result = sol.insert(intervals, new_interval)
        status = "PASS" if result == expected else "FAIL"
        print(f"{status} | intervals={intervals!r}, new={new_interval!r} => {result}")

File: solutions/057-insert-interval/solution.go
package main

import "fmt"

func insert(intervals [][]int, newInterval []int) [][]int {
	result := make([][]int, 0, len(intervals)+1)
	i, n := 0, len(intervals)

	for i < n && intervals[i][1] < newInterval[0] {
		result = append(result, intervals[i])
		i++
	}

	for i < n && intervals[i][0] <= newInterval[1] {
		if intervals[i][0] < newInterval[0] {
			newInterval[0] = intervals[i][0]
		}
		if intervals[i][1] > newInterval[1] {
			newInterval[1] = intervals[i][1]
		}
		i++
	}
	result = append(result, newInterval)

	for i < n {
		result = append(result, intervals[i])
		i++
	}
	return result
}

func equal(a, b [][]int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if len(a[i]) != 2 || len(b[i]) != 2 || a[i][0] != b[i][0] || a[i][1] != b[i][1] {
			return false
		}
	}
	return true
}

func main() {
	tests := []struct {
		intervals    [][]int
		newInterval  []int
		expected     [][]int
	}{
		{[][]int{{1, 3}, {6, 9}}, []int{2, 5}, [][]int{{1, 5}, {6, 9}}},
		{
			[][]int{{1, 2}, {3, 5}, {6, 7}, {8, 10}, {12, 16}},
			[]int{4, 8},
			[][]int{{1, 2}, {3, 10}, {12, 16}},
		},
		{[][]int{}, []int{5, 7}, [][]int{{5, 7}}},
		{[][]int{{1, 5}}, []int{2, 3}, [][]int{{1, 5}}},
	}

	for _, tc := range tests {
		newInterval := []int{tc.newInterval[0], tc.newInterval[1]}
		result := insert(tc.intervals, newInterval)
		status := "PASS"
		if !equal(result, tc.expected) {
			status = "FAIL"
		}
		fmt.Printf("%s | intervals=%v, new=%v => %v\n", status, tc.intervals, tc.newInterval, result)
	}
}
